File: scripts/extract_video_frames.m
#import <Foundation/Foundation.h>
#import <AVFoundation/AVFoundation.h>
#import <ImageIO/ImageIO.h>
#import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc < 3) {
            fprintf(stderr, "Usage: extract_video_frames <video> <output-dir> [frame-count]\n");
            return 1;
        }

        NSString *inputPath = [NSString stringWithUTF8String:argv[1]];
        NSString *outputPath = [NSString stringWithUTF8String:argv[2]];
        NSInteger frameCount = argc >= 4 ? MAX(2, atoi(argv[3])) : 8;
        [[NSFileManager defaultManager] createDirectoryAtPath:outputPath
                                  withIntermediateDirectories:YES
                                                   attributes:nil
                                                        error:nil];

        AVURLAsset *asset = [AVURLAsset URLAssetWithURL:[NSURL fileURLWithPath:inputPath] options:nil];
        Float64 duration = CMTimeGetSeconds(asset.duration);
        if (!isfinite(duration) || duration <= 0) {
            fprintf(stderr, "Could not read video duration\n");
            return 2;
        }

        AVAssetImageGenerator *generator = [[AVAssetImageGenerator alloc] initWithAsset:asset];
        generator.appliesPreferredTrackTransform = YES;
        generator.requestedTimeToleranceBefore = kCMTimeZero;
        generator.requestedTimeToleranceAfter = kCMTimeZero;
        NSString *stem = [[inputPath lastPathComponent] stringByDeletingPathExtension];

        for (NSInteger index = 0; index < frameCount; index++) {
            double fraction = ((double)index + 0.5) / (double)frameCount;
            double seconds = duration * fraction;
            CMTime time = CMTimeMakeWithSeconds(seconds, 600);
            NSError *error = nil;
            CGImageRef image = [generator copyCGImageAtTime:time actualTime:NULL error:&error];
            if (!image) {
                fprintf(stderr, "Frame %ld failed: %s\n", (long)index + 1,
                        error.localizedDescription.UTF8String);
                continue;
            }

            NSString *filename = [NSString stringWithFormat:@"%@-%02ld-%05.1fs.png",
                                  stem, (long)index + 1, seconds];
            NSURL *outputURL = [NSURL fileURLWithPath:[outputPath stringByAppendingPathComponent:filename]];
            CGImageDestinationRef destination = CGImageDestinationCreateWithURL(
                (__bridge CFURLRef)outputURL,
                (__bridge CFStringRef)UTTypePNG.identifier,
                1,
                NULL
            );
            CGImageDestinationAddImage(destination, image, NULL);
            CGImageDestinationFinalize(destination);
            CFRelease(destination);
            CGImageRelease(image);
            printf("%s\n", filename.UTF8String);
        }
    }
    return 0;
}

File: scripts/extract_video_frames.swift
import Foundation
import AVFoundation
import AppKit

guard CommandLine.arguments.count >= 3 else {
    fputs("Usage: extract_video_frames.swift <video> <output-dir> [frame-count]\n", stderr)
    exit(1)
}

let inputURL = URL(fileURLWithPath: CommandLine.arguments[1])
let outputURL = URL(fileURLWithPath: CommandLine.arguments[2], isDirectory: true)
let requestedCount = CommandLine.arguments.count >= 4 ? (Int(CommandLine.arguments[3]) ?? 8) : 8
let frameCount = max(2, requestedCount)

try FileManager.default.createDirectory(at: outputURL, withIntermediateDirectories: true)

let asset = AVURLAsset(url: inputURL)
let duration = CMTimeGetSeconds(asset.duration)
guard duration.isFinite, duration > 0 else {
    fputs("Could not read video duration\n", stderr)
    exit(2)
}

let generator = AVAssetImageGenerator(asset: asset)
generator.appliesPreferredTrackTransform = true
generator.requestedTimeToleranceBefore = .zero
generator.requestedTimeToleranceAfter = .zero

let stem = inputURL.deletingPathExtension().lastPathComponent
for index in 0..<frameCount {
    let fraction = (Double(index) + 0.5) / Double(frameCount)
    let seconds = duration * fraction
    let time = CMTime(seconds: seconds, preferredTimescale: 600)
    do {
        let cgImage = try generator.copyCGImage(at: time, actualTime: nil)
        let bitmap = NSBitmapImageRep(cgImage: cgImage)
        guard let png = bitmap.representation(using: .png, properties: [:]) else { continue }
        let filename = String(format: "%@-%02d-%05.1fs.png", stem, index + 1, seconds)
        try png.write(to: outputURL.appendingPathComponent(filename))
        print(filename)
    } catch {
        fputs("Frame \(index + 1) failed: \(error)\n", stderr)
    }
}

